### ISSUE TITLE
Validation dataloader has only 3 unique cases, but it is evaluated 250 iterations, repeatedly. Why?

### ISSUE THREAD

**Issue by hdnminh** [MIC-DKFZ/nnUNet#2023]
Hi Team,

In medical dataset UCL **prostate**, there are only **13** cases. When spliting data intro **dl_tr** and **dl_val**, their sizes are 9 cases and 3 cases, respectively.

When we run **self.train_step()** & **self.validation_step()** below, we pass through **Dataloader3Ds**: **self.dataloader_train** and **self.dataloader_val**.

1. When I check the length of these dataloader3D, they do not have their own length, so you limited it by the self.num_iterations_per_epoch of the loop. Why did you implement such like this?

```py
I debug it by:

count = 0
for _ in self.dataloader_train:
   count = count + 1
print(count)

--> This loop never ends.
```

3. Because self.dataloader_train has only 3 unique cases. So each of all 250 num_iterations_per_epoch shares the same cases, repeatedly. I don't know why you do that. We have 3 cases to run evaluation, so we just loop one of 3 cases once. Could you explain more, please?

The code [here](https://github.com/MIC-DKFZ/nnUNet/blob/95e74d8f50fd84c570820917f31ad9c72d30910b/nnunetv2/training/nnUNetTrainer/nnUNetTrainer.py#L1279)

```py
class nnUNetTrainer
    def run_training(self):
        self.on_train_start()# 

        for epoch in range(self.current_epoch, self.num_epochs):
            self.on_epoch_start()

            self.on_train_epoch_start()
            train_outputs = []
            
            for batch_id in range(self.num_iterations_per_epoch): # 250
                train_outputs.append(self.train_step(next(self.dataloader_train)))    <----------
            self.on_train_epoch_end(train_outputs)

            with torch.no_grad():
                self.on_validation_epoch_start()
                val_outputs = []
                for batch_id in range(self.num_val_iterations_per_epoch):
                    val_outputs.append(self.validation_step(next(self.dataloader_val)))  <----------
                self.on_validation_epoch_end(val_outputs)

            self.on_epoch_end()

        self.on_train_end()

```

**Comment by hdnminh** [comment]
Hi @Lars-Kraemer 
Can you help me?

**Comment by hdnminh** [comment]
Hi @Lars-Kraemer,
please help meee

**Comment by Lars-Kraemer** [comment]
Hey @hdnminh,

Please avoid spamming, we will try to resolve any issues as quickly as possible.
Regarding your questions:
nnUNet takes a fixed number of steps for training and validation independent of the dataset (250 steps for training and 50 steps for validation).
Firstly nnUnet does not work image base, instead it works patch-based. 
During training these patches are not sampled image-wise, instead, it samples patches class-based. Simplified: Select a class; then select an image that contains this class; then select a patch that contains the class from this image. This ensures a sufficient representation of all classes during training and handles the problem of class imbalance. However, this also means that several patches can be sampled from one image per epoch. The value of 250 steps is in itself arbitrary but has been experimentally proven to be appropriate.
During validation, nnUNet does not iterate over all images as this is very inefficient for most datasets, and for medical datasets images often contain a lot of background which says little about the performance of the model.  Instead, patches based on the classes in the dataset are selected again in order to be able to better evaluate the performance of the model across all classes, this is called pseudo-validation. 
At the end of the training, a real validation is performed again by iterating once completely over all images in the validation set. The results are saved in fold_x/validation/summary.json.

Best,
Lars

**Comment by hdnminh** [comment]
Thank you so much. I apologize for rushing you. I will try to remain calm next time.

I wish you all the best with your plans.
